common/achievement/get.ts

```diff
@@ -156,37 +156,32 @@ const assembleAchievementData = async (userAchievements: achievements_with_templ
-    ) {
-        const dataAggregationKeys = Object.keys(userAchievements[currentAchievementIndex].template.conditionDataAggregations as Prisma.JsonObject);
+    if (userAchievements[currentAchievementIndex].template.type === AchievementType.STREAK) {
         const evaluationResult = await isAchievementConditionMet(userAchievements[currentAchievementIndex]);
-        if (evaluationResult) {
-            currentValue = dataAggregationKeys.map((key) => evaluationResult.resultObject[key]).reduce((a, b) => a + b, 0);
-            maxValue =
-                userAchievements[currentAchievementIndex].template.type === AchievementType.STREAK
-                    ? userAchievements[currentAchievementIndex].recordValue !== null && userAchievements[currentAchievementIndex].recordValue > currentValue
-                        ? userAchievements[currentAchievementIndex].recordValue
-                        : currentValue
-                    : dataAggregationKeys
-                          .map((key) => {
-                              // TODO: check if we can remove valueToAchieve
-                              return Number((userAchievements[currentAchievementIndex].template.conditionDataAggregations as any)[key].valueToAchieve || 0);
-                          })
-                          .reduce((a, b) => a + b, 0);
-            if (
-                userAchievements[currentAchievementIndex].template.type === AchievementType.STREAK &&
-                currentValue < maxValue &&
-                userAchievements[currentAchievementIndex].achievedAt
-            ) {
-                await prisma.user_achievement.update({
-                    where: { id: userAchievements[currentAchievementIndex].id },
-                    data: { achievedAt: null, isSeen: false },
-                });
-            }
+        currentValue = evaluationResult.aggregationResult;
+        maxValue = userAchievements[currentAchievementIndex].recordValue;
+
+        // Streaks are usually invalidated if a user does not show up for a certain meeting / course / etc.
+        // This means we don't have an event to reevaluate the streak naturally.
+        // There are two ways to solve this:
+        // 1. We could reevaluate the streak every time the user opens the app.
+        // 2. We could create a cronjob that reevaluates all streaks every now and then.
+        // As the cron would be more complex and less efficient, we chose the first option, which is implemented here:
+        if (evaluationResult.shouldInvalidateStreak) {
+            await prisma.user_achievement.update({
+                where: { id: userAchievements[currentAchievementIndex].id },
+                data: { achievedAt: null, isSeen: false },
+            });
         }
     }
+    if (userAchievements[currentAchievementIndex].template.type === AchievementType.TIERED) {
+        const evaluationResult = await isAchievementConditionMet(userAchievements[currentAchievementIndex]);
```

Comment:
Why do we evaluate a second time here?

Reply:
Theoretically, parts of the function are a big switch, like:
```
switch(type)
  case Streak:
     evaluate();
     setCurrent;
     setMax;
  case Tiered:
     evaluate();
     setCurrent;
     setMax;
  default:
     setCurrent;
     setMax;
```

But to make it more readable i've implemented it like this:
```
setDefaultCurrent;
setDefaultMax;
if (isStreak)
     evaluate();
     setCurrent;
     setMax;
}
if (isTiered)
     evaluate();
     setCurrent;
     setMax;
}
```

So, it's actually not called a second time, but only for the types we need the re-evaluation.